ui/raidboss/popup-text.ts

```diff
@@ -190,101 +314,168 @@ class TriggerOutputProxy {
     });
   }
 
-  getReplacement(template, params, name) {
+  getReplacement(
+      // Can't use optional modifier for this arg since the others aren't optional
+      template: { [lang: string]: unknown } | undefined,
+      params: TriggerParams,
+      name: string,
+      id: string): string | undefined {
     if (!template)
-      return null;
-    if (typeof template === 'object') {
-      if (this.displayLang in template)
-        template = template[this.displayLang];
-      else
-        template = template['en'];
-    }
-    if (typeof template !== 'string') {
-      console.error(`Trigger ${this.trigger.id} has invalid outputString ${name}.`);
-      return null;
+      return;
+
+    const value = template[this.displayLang] ?? template['en'];
+
+    if (typeof value !== 'string') {
+      console.error(`Trigger ${id} has invalid outputString ${name}.`);
+      return;
     }
 
-    return template.replace(/\${\s*([^}\s]+)\s*}/g, (fullMatch, key) => {
+    return value.replace(/\${\s*([^}\s]+)\s*}/g, (_fullMatch: string, key: string) => {
       if (params && key in params) {
         const str = params[key];
         if (typeof str !== 'string' && typeof str !== 'number') {
-          console.error(`Trigger ${this.trigger.id} has non-string param value ${key}.`);
+          console.error(`Trigger ${id} has non-string param value ${key}.`);
           return this.unknownValue;
         }
         return str;
       }
-      console.error(`Trigger ${this.trigger.id} can't replace ${key} in ${template}.`);
+      console.error(`Trigger ${id} can't replace ${key} in ${JSON.stringify(template)}.`);
       return this.unknownValue;
     });
   }
+
+  static makeOutput(
+      trigger: ProcessedTrigger,
+      displayLang: Lang,
+      perTriggerAutoConfig?: PerTriggerAutoConfig): Output {
+    // `Output` is the common type used for the trigger data interface to support arbitrary
+    // string keys and always returns a string. However, TypeScript doesn't have good support
+    // for the Proxy representing this structure so we need to cast Proxy => unknown => Output
+    return new TriggerOutputProxy(trigger, displayLang,
+        perTriggerAutoConfig) as unknown as Output;
+  }
+}
+
+export type RaidbossTriggerField =
+  TriggerField<RaidbossData, TriggerOutput<RaidbossData, MatchesAny>>;
+export type RaidbossTriggerOutput = TriggerOutput<RaidbossData, MatchesAny>;
+
+const defaultOutput = TriggerOutputProxy.makeOutput({}, 'en');
+
+export abstract class TriggerHelper {
```

Comment:
What does `abstract` get you over having fields be required?

Reply:
If it's not `abstract` then it needs to initialize every property that's not optional as part of the constructor to some sort of default, which is completely pointless.

I didn't realize that an interface can define a field, and then an implementation can implement that field as a getter, though, that seems better. See latest commit.